compose/foundation/foundation/src/skikoMain/kotlin/androidx/compose/foundation/Draggable.skiko.kt

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.foundation
+
+import androidx.compose.foundation.gestures.awaitPointerSlopOrCancellation
+import androidx.compose.foundation.gestures.drag
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberUpdatedState
+import androidx.compose.runtime.setValue
+import androidx.compose.runtime.snapshotFlow
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.composed
+import androidx.compose.ui.geometry.Offset
+import androidx.compose.ui.input.pointer.AwaitPointerEventScope
+import androidx.compose.ui.input.pointer.PointerEvent
+import androidx.compose.ui.input.pointer.PointerEventType
+import androidx.compose.ui.input.pointer.PointerInputChange
+import androidx.compose.ui.input.pointer.PointerKeyboardModifiers
+import androidx.compose.ui.input.pointer.PointerType
+import androidx.compose.ui.input.pointer.changedToUpIgnoreConsumed
+import androidx.compose.ui.input.pointer.pointerInput
+import androidx.compose.ui.input.pointer.positionChange
+import androidx.compose.ui.platform.LocalWindowInfo
+import androidx.compose.ui.util.fastAll
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.currentCoroutineContext
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+
+@ExperimentalFoundationApi
+data class DragChange(
+    val offset: Offset = Offset.Zero,
+    val previousKeyboardModifiers: PointerKeyboardModifiers,
+    val currentKeyboardModifiers: PointerKeyboardModifiers
+) {
+    fun keyboardModifierChanged(): Boolean =
+        previousKeyboardModifiers != currentKeyboardModifiers
+}
+
+@ExperimentalFoundationApi
+    /**
+     * @param onDrag - receives an instance of [DragChange]  for every pointer position change when dragging or
+     * when [PointerKeyboardModifiers] state changes after drag started and before it ends.
+     */
+fun Modifier.onDrag(
```

Comment:
In the last commit I changed it like this:

moved the file into .gestures package in skikoMain (to match the commonMain of foundation module)

there two functions:
```kotlin
suspend fun PointerInputScope.detectDragGestures(
    filter: PointerFilter.() -> Unit, // no default value here, otherwise there're 2 possible resolutions: new and old funs
    onDragStart: (Offset) -> Unit = {},
    onDragCancel: () -> Unit = {},
    onDragEnd: () -> Unit = {},
    onDrag: (Offset) -> Unit
) ...
// PointerFilter.Default can be used as a default value. But I would I agree that it's less convenient. Alternative is different name for this new function.

fun Modifier.onDragGesture(
    enabled: Boolean = true,
    filter: PointerFilter.() -> Unit = PointerFilter.Default,
    onDragStart: (Offset) -> Unit = {},
    onDragCancel: () -> Unit = {},
    onDragEnd: () -> Unit = {},
    onDrag: (Offset) -> Unit
): Modifier ... 
```

the 2nd one simply calls the first one. 

Pros of having such 2 functions:
- `Modifier.onDragGesture` hides some small but still boilerplate (pointerInput)
- `PointerInputScope.detectDragGestures` is quite similar to the existing detectDragGestures api (new onDrag callback doesn't have PointerInputChange). Therefore this function will simplify the migration from the old api

Cons:
- more ways to handle drag gestures

I still like `Modifier.onDragGesture` more. If we need to remove 1 of them, I would prefer to remove the extension `detectDragGestures`.  But I mostly look at this question from the perspective of a fresh new project. 
What do you think?

Reply:
I like `onDragGesture` better. It is cleaner, and users don't have to bother restarting the coroutine, if they capture some parameter inside the callback:
```
@Composable Widget(param: String) {
  pointerInput(param) {
    detectDragGestures {
       param
    }
  }
}
```
`param` will be captured from the first composition, and we have to restart the coroutine if it is changed, or use `rememberUpdatedState`. If we create lambda inside Composable, not inside coroutine, it will be automatically capture the new `param` on the recomposition.

But we need to discuss this with Google (not now, when we will upstream changes, and will be removing Experimental annotation), and decide which function should stay, and which should be deprecated. Maybe we don't know something, and the coroutine version is better.

One of the advantages of coroutine functions is we can build complex gesture detectors: wait for a long click, a drag, and a double click. But it seems if function already has an event loop inside, we can't easily combine it with another functions. So this advantage doesn't applicable to `detectDragGestures`.

Reply:
A question:

``` kotlin
pointerInput(param) { ... }
```
If `param` changes, it will replace an old pointInput with a new one. 
If this happens when drag in progress, it behaves not really nice - all old callbacks stop working, and new callbacks don't receive any events, since new pointInput suspends on the first awaitPressEvent (new pointerInput doesn't know anything about previous gesture and its progress).

This means that users need to take care about what keys they use for pointerInput to avoid such problems (when gesture is in progress and recomposition happens that leads to new pointerInput).
Is my understanding correct?

Reply:
Yes. It seems the only valid option when we use pointerInput is to use rememberUpdatedState